pkg/services/pluginsintegration/clientmiddleware/caching_middleware_test.go

```diff
@@ -0,0 +1,175 @@
+package clientmiddleware
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"testing"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/plugins/manager/client/clienttest"
+	"github.com/grafana/grafana/pkg/services/caching"
+	"github.com/grafana/grafana/pkg/services/contexthandler"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCachingMiddleware(t *testing.T) {
+	t.Run("When QueryData is called", func(t *testing.T) {
+		req, err := http.NewRequest(http.MethodGet, "/query", nil)
+		require.NoError(t, err)
+
+		cs := caching.NewFakeOSSCachingService(t)
+		cdt := clienttest.NewClientDecoratorTest(t,
+			clienttest.WithReqContext(req, &user.SignedInUser{}),
+			clienttest.WithMiddlewares(NewCachingMiddleware(cs)),
+		)
+
+		jsonDataMap := map[string]interface{}{}
+		jsonDataBytes, err := json.Marshal(&jsonDataMap)
+		require.NoError(t, err)
+
+		pluginCtx := backend.PluginContext{
+			DataSourceInstanceSettings: &backend.DataSourceInstanceSettings{
+				JSONData: jsonDataBytes,
+			},
+		}
+
+		// Populated by clienttest.WithReqContext
+		reqCtx := contexthandler.FromContext(req.Context())
+		require.NotNil(t, reqCtx)
+
+		qdr := &backend.QueryDataRequest{
+			PluginContext: pluginCtx,
+		}
+
+		// Track whether the update cache fn was called, depending on what the response headers are in the cache request
+		var updateCacheCalled bool
+		dataResponse := caching.CachedQueryDataResponse{
+			Response: &backend.QueryDataResponse{},
+			UpdateCacheFn: func(ctx context.Context, qdr *backend.QueryDataResponse) {
+				updateCacheCalled = true
+			},
+		}
+
+		call := cs.On("HandleQueryRequest", req.Context(), qdr).Return(dataResponse)
+
+		t.Run("If cache returns a hit, no queries are issued", func(t *testing.T) {
+			t.Cleanup(func() {
+				updateCacheCalled = false
+			})
+
+			call.Run(func(args mock.Arguments) {
+				reqCtx.Resp.Header().Set(caching.XCacheHeader, caching.StatusHit)
+			})
+
+			resp, err := cdt.Decorator.QueryData(req.Context(), qdr)
+			assert.NoError(t, err)
+			// Cache service is called once
+			cs.AssertNumberOfCalls(t, "HandleQueryRequest", 1)
+			// Equals the mocked response
+			assert.NotNil(t, resp)
+			assert.Equal(t, dataResponse.Response, resp)
+			// Cache was not updated by the middleware
+			assert.False(t, updateCacheCalled)
+		})
+
+		t.Run("If cache returns a miss, queries are issued and the update cache function is called", func(t *testing.T) {
+			t.Cleanup(func() {
+				updateCacheCalled = false
+			})
+
+			call.Run(func(args mock.Arguments) {
+				reqCtx.Resp.Header().Set(caching.XCacheHeader, caching.StatusMiss)
+			})
+
+			resp, err := cdt.Decorator.QueryData(req.Context(), qdr)
+			assert.NoError(t, err)
+			// Cache service is called again
+			cs.AssertNumberOfCalls(t, "HandleQueryRequest", 2)
```

Comment:
this check depends on the above test, therefore it fails if the test runs alone:
<details>
<summary>test output</summary>

```
~/go/src/github.com/grafana/grafana: go test -v ./pkg/services/pluginsintegration/clientmiddleware/... -run TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called
=== RUN   TestCachingMiddleware
=== RUN   TestCachingMiddleware/When_QueryData_is_called
=== RUN   TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called
    caching_middleware_test.go:91: 
                Error Trace:    /Users/josefk/go/src/github.com/grafana/grafana/pkg/services/pluginsintegration/clientmiddleware/caching_middleware_test.go:91
                Error:          Not equal: 
                                expected: 2
                                actual  : 1
                Test:           TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called
                Messages:       Expected number of calls (2) does not match the actual number of calls (1).
=== NAME  TestCachingMiddleware/When_QueryData_is_called
    caching_service_mock.go:63: PASS:   HandleQueryRequest(*context.valueCtx,*backend.QueryDataRequest)
--- FAIL: TestCachingMiddleware (0.00s)
    --- FAIL: TestCachingMiddleware/When_QueryData_is_called (0.00s)
        --- FAIL: TestCachingMiddleware/When_QueryData_is_called/If_cache_returns_a_miss,_queries_are_issued_and_the_update_cache_function_is_called (0.00s)
FAIL
FAIL    github.com/grafana/grafana/pkg/services/pluginsintegration/clientmiddleware     0.369s
FAIL
```
</details>

it would be better if the tests were independent (create a new fake service for each test maybe?)

Reply:
Good catch. I just made an update to remove mockery and Reset my fake service between tests, so this is fixed now. Just verified they run independently. [1bb82aa](https://github.com/grafana/grafana/pull/65616/commits/1bb82aa25e5da0d997a11adbf0341e6454bbea68)